cmd/agent/main.go

```diff
@@ -1,29 +1,238 @@
+	namespace        = flag.String("namespace", "", "The namespace to add as header to log events")
+	endpoint         = flag.String("endpoint", "", "The endpoint name to add as header to log events")
+
+	// probing flags
+	readinessProbeTimeout = flag.Duration("probe-period", -1, "run readiness probe with given timeout")
+	// This creates an abstract socket instead of an actual file.
+	unixSocketPath = "@/kfserving/agent.sock"
 )
 
+type config struct {
+	ContainerConcurrency   int    `split_words:"true" required:"true"`
+	QueueServingPort       int    `split_words:"true" required:"true"`
+	UserPort               int    `split_words:"true" required:"true"`
+	RevisionTimeoutSeconds int    `split_words:"true" required:"true"`
+	ServingReadinessProbe  string `split_words:"true" required:"true"`
+	// Logging configuration
+	ServingLoggingConfig         string `split_words:"true" required:"true"`
+	ServingLoggingLevel          string `split_words:"true" required:"true"`
+	ServingRequestLogTemplate    string `split_words:"true"` // optional
+	ServingEnableRequestLog      bool   `split_words:"true"` // optional
+	ServingEnableProbeRequestLog bool   `split_words:"true"` // optional
+}
+
+type loggerArgs struct {
+	loggerType       v1beta1.LoggerType
+	logUrl           *url.URL
+	sourceUrl        *url.URL
+	inferenceService string
+	namespace        string
+	endpoint         string
+}
+
 func main() {
 	flag.Parse()
-	logf.SetLogger(logf.ZapLogger(false))
-	log.Info("Initializing model agent with", "config-dir", configDir, "model-dir", modelDir)
+	// Parse the environment.
+	var env config
+	if err := envconfig.Process("", &env); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+	// If this is set, we run as a standalone binary to probe the queue-proxy.
+	if *readinessProbeTimeout >= 0 {
+		// Use a unix socket rather than TCP to avoid going via entire TCP stack
+		// when we're actually in the same container.
+		transport := http.DefaultTransport.(*http.Transport).Clone()
+		transport.DialContext = func(_ context.Context, _, _ string) (net.Conn, error) {
+			return net.Dial("unix", unixSocketPath)
+		}
+
+		os.Exit(standaloneProbeMain(*readinessProbeTimeout, transport, *port))
+	}
+	logger, _ := pkglogging.NewLogger(env.ServingLoggingConfig, env.ServingLoggingLevel)
+
+	if *enablePuller {
+		startModelPuller(logger)
+	}
+
+	var loggerArgs *loggerArgs
+	if *logUrl != "" {
+		loggerArgs = startLogger(*workers, logger)
+	}
+
+	healthState := &health.State{}
+	ctx := signals.NewContext()
+	// Setup probe to run for checking user-application healthiness.
+	probe := buildProbe(logger, env.ServingReadinessProbe, *componentPort)
+	mainServer := buildServer(ctx, *port, *componentPort, loggerArgs, healthState, probe, logger)
+	servers := map[string]*http.Server{
+		"main": mainServer,
+	}
+	errCh := make(chan error)
+	listenCh := make(chan struct{})
```

Comment:
Should we reference the copy of code from https://github.com/knative/serving/blob/master/cmd/queue/main.go ?

Reply:
sure